Let KINSOL callbacks conform to our new standard for callback errors.

bangerth opened the pull request:
As laid out in #15112 and documented in #15146.

I broke this into a number of individual commit that might be easier to understand in separation:
* fcb062c is the one that actually converts user callback exceptions into the return codes KINSOL wants to see (in the `call_and_possibly_capture_exception()` function), and consequently changes the interface of the `std::function` callbacks.
* 91fbf68 converts the tests. Take a particular look at `kinsol_06.cc`, the test introduced in #15172 that illustrates how KINSOL deals with recoverable errors.
* 6dba3d3 converts step-77, shrinking the size of the callbacks by about 1/3 :-)
* cfc5539 cleans up a wart in the first of the commits mentioned above: I introduce a variable `std::exception_ptr pending_exception;` that I needed to make `public` because it is referenced in several free functions. Each of these free functions is used in only one place, so I'm turning these free functions into lambdas -- and that allows me to make the member variable `private`. This is because the following is apparently ok:
```
  class X {
    private:
      int i;
      void f();
  };

  void X::f() {
    auto callback = []() {
     X x;   // create an object in the 'callback' lambda
     return x.i; // access its private member function -- can do because we are *in* X right now
    };
  }
```

I should note that this is backward compatible at least for the case where user callbacks return zero (=no error). That's because you can assign a function that returns something to a `std::function<void(...)>`. The returned value will then simply be ignored. Once I'm done with #15112, I will write a changelog entry that documents all of this. In the meantime, my suspicion is that 100% of user callbacks actually return zero, and so there is no actual change; the only change is if callbacks return nonzero codes, which will now be ignored.

All of this passes all 27 SUNDIALS tests for me.

@sebproell Since you know SUNDIALS, I thought you might be interested in looking at this. I will do something similar for the NOX as well as for the other SUNDIALS interfaces.
@stefanozampini I would do something similar for the PETSc callbacks.

bangerth commented:
@singima Regarding your patch #15160, this is backward compatible and so the two don't step on each other. I will adjust your new class in the same way once this is merged and I've taken care of the NOX interfaces.

stefanozampini commented:
@bangerth I can take care of the PETSc interface once you are done with kinsol and nox.

bangerth commented:
@stefanozampini  I will certainly not argue with that offer :-)

bangerth commented:
The one failing check is the unrelated test case that's currently broken in master.

sebproell commented on include/deal.II/sundials/kinsol.h:
```diff
@@ -423,29 +431,29 @@ namespace SUNDIALS
+     * can throw exceptions of type RecoverableUserCallbackError.
      */
-    std::function<int(const VectorType &src, VectorType &dst)> residual;
+    std::function<void(const VectorType &src, VectorType &dst)> residual;
```
Is this really backwards-compatible? I don't think you can assign existing  callbacks that look like 

```c++
[](const VectorType &src, VectorType &dst) { 
  // some code
  return 0;
}
```
due to the (deduced) return type being int instead of void.

sebproell replied:
Nvm, my quick check in compiler explorer which seemed to indicate this is not possible was flawed for another reason... I can confirm this works ;)

bangerth replied:
:-) `std::function` really does not care very much what the object is that it gets to represent, as long as it can be called with the arguments given when one calls `std::function::operator()`, and as long as the result of the called thing can be assigned to the return code of the `std::function` object (which I changed from `int` to `void` -- anything can be assigned to `void`, or put differently: the return value of the thing the `std::function` stores is simply ignored).

All of this means that this code snippet compiles and runs just fine:
```
#include <functional>

int f(int);

int main ()
{
  std::function<void(double)> fun = &f;    // note mismatch in return type and argument
  fun(1.234);   // ok -- fun.operator() takes a 'double' as argument, which is implicitly converted to 'int' when calling 'f'
}
```

peterrum replied:
Independent of if it compiles or no not, we probably need a change-log entry, since if the return value is just ignored this means that means the behavior will be different.

tjhei reviewed (approved):
Can you add a changelog entry? Otherwise this looks good to go to me.

bangerth commented:
OK, I relent :-) By popular demand, I have added a changelog entry that I will simply adapt for every changed interface.

The patch itself tested cleanly in case someone wants to merge without waiting for all testers to go through it all again. The only change since the last clean test is the new changelog.
